### PR TITLE
Cloud Integration Update

### LINKED ISSUE #3047: [Feature] Allow embedded file configs in values
### Problem Statement

It is possible to pass a configuration via helm values without using "extraObjects" or creating out-of-band secrets/configmaps.
Kubecost helm chart does not support this for many of the configuration items we require for integrating with cloud billing/object-storage/etc.

Benefits for embedding the configs:
1. Simplicity
2. Less mistakes (foot guns)
3. Easier to support for future users that take over the role
4. single file "advanced" configs

### Solution Description

Example:
https://github.com/bitnami/charts/blob/main/bitnami/thanos/templates/objstore-secret.yaml#L14
values file is simply:
```yaml
objstoreConfig: |-
  type: S3
  config:
    bucket: my-bucket
    region: eu-central-1
    endpoint: s3.amazonaws.com
```

### Alternatives

_No response_

### Additional Context

_No response_

### Troubleshooting

- [X] I have read and followed the [issue guidelines](https://github.com/kubecost/cost-analyzer-helm-chart/blob/develop/ISSUE_GUIDELINES.md) and this is a feature request only for the Helm chart.
- [X] I have searched other issues in this repository and mine is not recorded.

### PR DESCRIPTION
**Pull request by chipzoller**
Signed-off-by: chipzoller <chipzoller@gmail.com>

## What does this PR change?
Two fixes here:
1. Fixes a bug where cloud billing via KubecostProductConfigs was not honored.
2. 
Adds a new key under `kubecostProductConfigs` called `cloudIntegrationJSON`, the value of which is a block scalar in JSON form with the cloud integration contents. This is provided as an alternative for users who would rather define the cloud integration contents in-line to the values file rather than precreating a Secret with similar contents. This new value is mutually exclusive with the existing `cloudIntegrationSecret` and only one may be specified. If both are specified, installation will fail.

This PR also adds a second check for upgrade protection which will also produce a failure if users have set individual fields under kubecostProductConfigs with cloud-specific information previously used to create cloud integration info.

## Does this PR rely on any other PRs?

No but is related to #3048

## How does this PR impact users? (This is the kind of thing that goes in release notes!)

Allows users to define the cloud integration contents in their values file rather than being forced to precreate a Secret and provide that value to this chart's installation process.

## Links to Issues or tickets this PR addresses or fixes

Closes #3047

## What risks are associated with merging this PR? What is required to fully test this PR?

There could be issues when using either when templating, or the encoded value could trip up Kubecost.

## How was this PR tested?

Helm templating in both combinations and comparing results.

### Specifying both cloudIntegrationSecret and cloudIntegrationJSON

```sh
$ helm template kubecost cost-analyzer/
Error: execution error at (cost-analyzer/templates/NOTES.txt:3:4): cloudIntegrationSecret and cloudIntegrationJSON are mutually exclusive. Please specify only one.
```

### Rendered Secret named cloud-integration

```yaml
apiVersion: v1
kind: Secret
type: Opaque
metadata:
  name: cloud-integration
  namespace: default
  labels:
    app.kubernetes.io/name: cost-analyzer
    helm.sh/chart: cost-analyzer-2.0.0
    app.kubernetes.io/instance: kubecost
    app.kubernetes.io/managed-by: Helm
    app: cost-analyzer
data:
  cloud-integration.json: eyAgImF3cyI6IFsgICAgeyAgICAgICJhdGhlbmFCdWNrZXROYW1lIjogInMzOi8vQVdTX2Nsb3VkX2ludGVncmF0aW9uX2F0aGVuYUJ1Y2tldE5hbWUiLCAgICAgICJhdGhlbmFSZWdpb24iOiAiQVdTX2Nsb3VkX2ludGVncmF0aW9uX2F0aGVuYVJlZ2lvbiIsICAgICAgImF0aGVuYURhdGFiYXNlIjogIkFXU19jbG91ZF9pbnRlZ3JhdGlvbl9hdGhlbmFEYXRhYmFzZSIsICAgICAgImF0aGVuYVRhYmxlIjogIkFXU19jbG91ZF9pbnRlZ3JhdGlvbl9hdGhlbmFCdWNrZXROYW1lIiwgICAgICAicHJvamVjdElEIjogIkFXU19jbG91ZF9pbnRlZ3JhdGlvbl9hdGhlbmFfcHJvamVjdElEIiwgICAgICAic2VydmljZUtleU5hbWUiOiAiQVdTX2Nsb3VkX2ludGVncmF0aW9uX2F0aGVuYV9zZXJ2aWNlS2V5TmFtZSIsICAgICAgInNlcnZpY2VLZXlTZWNyZXQiOiAiQVdTX2Nsb3VkX2ludGVncmF0aW9uX2F0aGVuYV9zZXJ2aWNlS2V5U2VjcmV0IiAgICB9ICBdfQ==
```

### Contents of Secret

Input values file

```sh
$ yq e .kubecostProductConfigs.cloudIntegrationJSON cost-analyzer/values.yaml | jq
{
  "aws": [
    {
      "athenaBucketName": "s3://AWS_cloud_integration_athenaBucketName",
      "athenaRegion": "AWS_cloud_integration_athenaRegion",
      "athenaDatabase": "AWS_cloud_integration_athenaDatabase",
      "athenaTable": "AWS_cloud_integration_athenaBucketName",
      "projectID": "AWS_cloud_integration_athena_projectID",
      "serviceKeyName": "AWS_cloud_integration_athena_serviceKeyName",
      "serviceKeySecret": "AWS_cloud_integration_athena_serviceKeySecret"
    }
  ]
}
```

Decoded contents of the templatized Secret

```sh
$ helm template kubecost cost-analyzer/ -s templates/cloud-integration-secret.yaml | yq .data.\"cloud-integration.json\" | base64 -d | jq
{
  "aws": [
    {
      "athenaBucketName": "s3://AWS_cloud_integration_athenaBucketName",
      "athenaRegion": "AWS_cloud_integration_athenaRegion",
      "athenaDatabase": "AWS_cloud_integration_athenaDatabase",
      "athenaTable": "AWS_cloud_integration_athenaBucketName",
      "projectID": "AWS_cloud_integration_athena_projectID",
      "serviceKeyName": "AWS_cloud_integration_athena_serviceKeyName",
      "serviceKeySecret": "AWS_cloud_integration_athena_serviceKeySecret"
    }
  ]
}
```

## Have you made an update to documentation? If so, please provide the corresponding PR.

No

### CODE REVIEW

**Comment by chipzoller** [comment]
For those requested for review, I could use your help testing this to ensure Kubecost works properly after consuming the Secret since I'm dealing with test/junk data.

File: cost-analyzer/templates/aggregator-cloud-cost-deployment.yaml [hunk 1]
```diff
@@ -51,12 +51,21 @@ spec:
             defaultMode: 420
             secretName: {{ .Values.kubecostModel.federatedStorageConfigSecret }}
         {{- end }}
+        {{- if .Values.kubecostProductConfigs.cloudIntegrationSecret }}
         - name: cloud-integration
           secret:
             secretName: {{ .Values.kubecostProductConfigs.cloudIntegrationSecret }}
+            items:
+              - key: cloud-integration.json
+                path: cloud-integration.
```
Comment (ameijer) [inline cost-analyzer/templates/aggregator-cloud-cost-deployment.yaml]:
```suggestion
                path: cloud-integration.json
```

?

Comment (chipzoller) [reply cost-analyzer/templates/aggregator-cloud-cost-deployment.yaml]: Shit. Thank you very much, Alex. Fixed.

File: cost-analyzer/templates/NOTES.txt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 --------------------------------------------------
 {{- include "kubecostV2-preconditions" . -}}
+{{- include "cloudIntegrationSecretMutualExclusivityCheck" . -}}
```
Comment (ameijer) [inline cost-analyzer/templates/NOTES.txt]: more of a philosophical discussion here, but do you know when NOTES is rendered, chip? I can't find it in https://github.com/helm/helm/blob/release-3.0/pkg/releaseutil/kind_sorter.go#L27 . I would hate to have it rendered at the end and we blow the install up after the bad stuff has already been sent to API server

File: cost-analyzer/values.yaml [hunk 3]
```diff
@@ -2892,109 +2892,131 @@ costEventsAudit:
...
+  ## Specify an existing Kubernetes Secret holding the cloud integration information. This Secret must contain
+  ## a key with name `cloud-integration.json` and the contents must be in a specific format. It is expected
+  ## to exist in the release Namespace. This is mutually exclusive with cloudIntegrationJSON where only one must be defined.
+  # cloudIntegrationSecret: "cloud-integration"
+
+  ## Specify the cloud integration information in JSON form if pointing to an existing Secret is not desired or you'd rather
+  ## define the cloud integration information directly in the values file. This will result in a new Secret being created
+  ## named `cloud-integration` in the release Namespace. It is mutually exclusive with the cloudIntegrationSecret where only one must be defined.
+  # cloudIntegrationJSON: |-
+  #   {
```
Comment (ameijer) [inline cost-analyzer/values.yaml:3007]: would it make sense if we also added example configs for GCP, and Azure?

File: cost-analyzer/templates/NOTES.txt (hunk 2 above)
Comment (chipzoller) [reply cost-analyzer/templates/NOTES.txt]: All the resources are rendered internally before being parcelled out to the API server in the order described [here](https://helm.sh/docs/intro/using_helm/). If a failure is encountered, no resources will be submitted.

```sh
$ k get ns
NAME              STATUS   AGE
kube-system       Active   41h
default           Active   41h
kube-public       Active   41h
kube-node-lease   Active   41h

$ helm install kubecost -n kubecost cost-analyzer/ --create-namespace
Error: INSTALLATION FAILED: execution error at (cost-analyzer/templates/NOTES.txt:3:4): cloudIntegrationSecret and cloudIntegrationJSON are mutually exclusive. Please specify only one.

$ k get ns
NAME              STATUS   AGE
kube-system       Active   41h
default           Active   41h
kube-public       Active   41h
kube-node-lease   Active   41h
```

File: cost-analyzer/values.yaml (hunk 3 above)
Comment (chipzoller) [reply cost-analyzer/values.yaml:3007]: That would be great, yes. Can you supply something that here, in a comment, so I can include? I know you are aware, but ensure there's nothing sensitive in those values :)

Comment (ameijer) [reply cost-analyzer/values.yaml:3007]:
```
"azure": [
	{
		"azureSubscriptionID": "my-subscription-id",
		"azureStorageAccount": "my-storage-account",
		"azureStorageAccessKey": "my-storage-access-key",
		"azureStorageContainer": "my-storage-container"
	}
],
"gcp": [
	{
		"projectID": "my-project-id",
		"billingDataDataset": "detailedbilling.my-billing-dataset",
		"key": {
			"type": "service_account",
			"project_id": "my-project-id",
			"private_key_id": "my-private-key-id",
			"private_key": "my-pem-encoded-private-key",
			"client_email": "my-service-account-name@my-project-id.iam.gserviceaccount.com",
			"client_id": "my-client-id",
			"auth_uri": "auth-uri",
			"token_uri": "token-uri",
			"auth_provider_x509_cert_url": "my-x509-provider-cert",
			"client_x509_cert_url": "my-x509-cert-url"
		}
	}
],
```

Comment (chipzoller) [reply cost-analyzer/values.yaml:3007]: Thank you. How about the AWS values? Are those correct and also complete?

Comment (ameijer) [reply cost-analyzer/values.yaml:3007]: yes they look good so I omitted those

Comment (chipzoller) [reply cost-analyzer/values.yaml:3007]: Awesome, thanks. Just pushed those up so have a look.

**Review by ameijer — APPROVED** [review]
lgtm

**Review by Sean-Holcomb — APPROVED** [review]

**Review by cliffcolvin — APPROVED** [review]
:1st_place_medal:

File: cost-analyzer/templates/_helpers.tpl [hunk 4]
```diff
@@ -93,6 +93,39 @@ Kubecost 2.0 preconditions
...
+          "athenaTable": "{{ .Values.kubecostProductConfigs.athenaTable }}",
+          "projectID": "{{ .Values.kubecostProductConfigs.athenaProjectID }}"
+          {{- if and ((.Values.kubecostProductConfigs).awsServiceKeyName) ((.Values.kubecostProductConfigs).awsServiceKeyPassword) }},
+          "serviceKeyName": "{{ .Values.kubecostProductConfigs.awsServiceKeyName }}",
+          "serviceKeySecret": "{{ .Values.kubecostProductConfigs.awsServiceKeyPassword }}"
+          {{- end }}
+      }
+    ]
+  }
+{{- end }}
```
Comment (thomasvn) [inline cost-analyzer/templates/_helpers.tpl:112]: @Sean-Holcomb Is this still a valid way of setting up your cloud integration? I've been told from a user that the only supported method now is view `cloud-integration.json`.

Comment (jessegoodier) [reply cost-analyzer/templates/_helpers.tpl:112]: This creates that secret via helm, fixing that problem.

We also fixed the UI to only show status of cloud-integration and not the old API.

Comment (thomasvn) [reply cost-analyzer/templates/_helpers.tpl:112]: @jessegoodier Awesome! We'll do the same for Azure & GCP in a separate PR?

**Pull request merged**